### PR TITLE
feat: use hive clusterdeployment for creating spoke clusters

### PR DESCRIPTION
**Pull request by TomerFi**
- Added feature for creating Spoke Clusters using Hive's [ClusterDeployment API](https://github.com/openshift/hive/blob/master/docs/using-hive.md#clusterdeployment).
- Modified the _clustergroup_ schema; added `clusterDeployments` for a group, used `clusterPools` as a template.
- Modified test cases testing `clusterDeployments` on the same _multiclustergroup_ with `clusterPools` and on its own.
- Multiple Commits in this PR make reviewing easier; these can be squashed together if required.
- If preferred, we can move the tests to a different PR to make this smaller.

Example configuration:
```yaml
  managedClusterGroups:
    - name: group-with-pool
      clusterPools:
        spokesPool:
          name: spokes-pool
          openshiftVersion: 4.14.12
          baseDomain: domain.example.com
          platform:
            aws:
              region: ap-southeast-2
          clusters:
          - spoke1
          - spoke2

    - name: group-with-deployments
      clusterDeployments:
        myFirstCustomSpoke:
          name: my-first-custom-spoke
          openshiftVersion: 4.14.12
          baseDomain: domain.example.com
          platform:
            azure:
              baseDomainResourceGroupName: dojo-dns-zones
              region: eastus
        mySecondCustomSpoke:
          name: my-second-custom-spoke
          openshiftVersion: 4.14.12
          baseDomain: domain.example.ca
          platform:
            azure:
              baseDomainResourceGroupName: dojo-dns-zones
              region: westca

    # all clusters in the following group, claims or deployments, are considered part of the same
    # group, same labeling mechanism, shared managed cluster set, etc.
    - name: group-with-both
      clusterPools:
        anotherSpokesPool:
          name: another-spokes-pool
          openshiftVersion: 4.14.12
          baseDomain: domain.example.com
          platform:
            aws:
              region: ap-southwest-1
          clusters:
            - another-spoke1
      clusterDeployments:
        anotherCustomSpoke:
          name: another-custom-spoke
          openshiftVersion: 4.14.12
          baseDomain: domain.example.com
          platform:
            aws:
              region: ap-southwest-3

```

### CODE REVIEW

**Review by r2dedios — APPROVED** [review]
LGTM ;)

File: acm/templates/provision/clusterdeployment.yaml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  name: {{ $deploymentName }}
+
+---
+apiVersion: hive.openshift.io/v1
+kind: ClusterDeployment
+metadata:
+  name: {{ $deploymentName }}
+  namespace: {{ $deploymentName }}
+  labels:
+    vendor: OpenShift
```
[the remark this answers is not in the document]
Comment (TomerFi) [reply acm/templates/provision/clusterdeployment.yaml:32]: good to know - thanks!

File: acm/templates/provision/clusterdeployment.yaml [hunk 2]
```diff
@@ -0,0 +1,71 @@
+{{- range .Values.clusterGroup.managedClusterGroups }}
+{{- $group := . }}
+
+{{- range $group.clusterDeployments}}
+{{ $cluster := . }}
```
[the remark this answers is not in the document]
Comment (TomerFi) [reply acm/templates/provision/clusterdeployment.yaml:5]: ohh good catch! we'll add it here.

File: acm/templates/provision/managedclusterset.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+apiVersion: cluster.open-cluster-management.io/v1beta2
+kind: ManagedClusterSet
+metadata:
+  annotations:
+    cluster.open-cluster-management.io/submariner-broker-ns: {{ .name }}-broker
+    argocd.argoproj.io/sync-options: SkipDryRunOnMissingResource=true
+  name: {{ .name }}
+spec:
+  clusterSelector:
+    selectorType: LegacyClusterSetLabel
```
[the remark this answers is not in the document]
Comment (TomerFi) [reply acm/templates/provision/managedclusterset.yaml:13]: @mbaldessari I did some investigation.

This [issue](https://github.com/open-cluster-management-io/ocm/issues/67) describes [adding](https://github.com/open-cluster-management-io/api/pull/174/) _v1beta2_ depercating _LegacyClusterSetLabel_ (previous default) and adding _ExclusiveClusterSetLabel_ (current default). Later on, _v1beta1_ was [removed](https://github.com/open-cluster-management-io/api/pull/266) which explains our issue.

Currently available options for _SelectorType_:
- _ExclusiveClusterSetLabel_, the default value, means that _ManagedCluster_ resources will be selected based on the label: `cluster.open-cluster-management.io/clusterset:<ManagedClusterSet Name>`.
- _LabelSelector_ requires speicying a [LabelSelector](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#LabelSelector) spec key specifying a custom label for selection.

Considering both _LegacyClusterSetLabel_ and _ExclusiveClusterSetLabel_ are the default for their respective versions, we should delete this spec for backward compatibility.

File: acm/templates/provision/clusterdeployment.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (TomerFi) [reply acm/templates/provision/clusterdeployment.yaml:40]: Resolved in https://github.com/validatedpatterns/common/pull/472/commits/6ae92539a04dca4a2f8914fd16b64cb174ef85d1.

File: acm/templates/provision/clusterdeployment.yaml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+      name: {{ $deploymentName }}-ssh-private-key
+    imageSetRef:
+      name: img{{ $cluster.openshiftVersion }}-multi-appsub
+  pullSecretRef:
+    name: {{ $deploymentName }}-pull-secret
+
+---
+apiVersion: cluster.open-cluster-management.io/v1
+kind: ManagedCluster
+metadata:
```
[the remark this answers is not in the document]
Comment (TomerFi) [reply acm/templates/provision/clusterdeployment.yaml:65]: Resolved in https://github.com/validatedpatterns/common/pull/472/commits/6ae92539a04dca4a2f8914fd16b64cb174ef85d1.

**Comment by TomerFi** [comment]
@mbaldessari, can you please take a look at the [failed workflow](https://github.com/validatedpatterns/common/actions/runs/8807243824/job/24173861151?pr=472)?

Tests are passing on my end. Based on the failure message, I tried aligning the `valueFiles` array identification, but the result was the same.

File: acm/templates/provision/clusterdeployment.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (mbaldessari) [reply acm/templates/provision/clusterdeployment.yaml:5]: Yeah let's do later it in another PR maybe

File: acm/templates/provision/managedclusterset.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (mbaldessari) [reply acm/templates/provision/managedclusterset.yaml:13]: Agree, it seems we should just drop this part for the time being

**Comment by mbaldessari** [comment]
> @mbaldessari, can you please take a look at the [failed workflow](https://github.com/validatedpatterns/common/actions/runs/8807243824/job/24173861151?pr=472)?
> 
> Tests are passing on my end. Based on the failure message, I tried aligning the `valueFiles` array identification, but the result was the same.


> @mbaldessari, can you please take a look at the [failed workflow](https://github.com/validatedpatterns/common/actions/runs/8807243824/job/24173861151?pr=472)?
> 
> Tests are passing on my end. Based on the failure message, I tried aligning the `valueFiles` array identification, but the result was the same.

Aye we can worry / fix this one up later on. We fixed a bug in there in the meantime I suspect that is why the test is failing. 

I did try this PR + the drop of legacyClusterSelector and for some reason the additional cluster could not be spun up by ACM. I might have done something wrong. I'll poke you next week about it if I can't figure it out

File: acm/templates/provision/managedclusterset.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (TomerFi) [reply acm/templates/provision/managedclusterset.yaml]: My pleasure!

[the remark this answers is not in the document]
Comment (TomerFi) [reply acm/templates/provision/managedclusterset.yaml]: Removed in https://github.com/validatedpatterns/common/pull/472/commits/c6ffd0ecdab72ff98f3cf89271ab4c5433c9f5bf.

**Comment by TomerFi** [comment]
@mbaldessari @r2dedios 
Following our session today, namespaces was added to clusterdeployment-related secrets in https://github.com/validatedpatterns/common/pull/472/commits/6cd4e85d5bf7955b172b935dd05cd360fd0c4443.

**Comment by mbaldessari** [comment]
Thanks!

**Comment by r2dedios** [comment]
Thanks @mbaldessari

**Comment by TomerFi** [comment]
Thank you @mbaldessari and @r2dedios !

File: acm/templates/provision/clusterdeployment.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (TomerFi) [reply acm/templates/provision/clusterdeployment.yaml:5]: #524

**Pull request merged**